core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,305 @@
+  /**
+   * Cache of the input streams, from the input stream id to the input stream.
+   */
+  private final Cache<Long, SeekableUnderFileInputStream> mUnderFileInputStreamCache;
+  /**
+   * Thread pool for asynchronously removing the expired input streams.
+   */
+  private final ExecutorService mRemovalThreadPool;
+
+  /**
+   * A listener to the input stream removal.
+   */
+  private RemovalListener<Long, SeekableUnderFileInputStream> mRemovalListener =
+      new RemovalListener<Long, SeekableUnderFileInputStream>() {
+        @Override
+        public void onRemoval(RemovalNotification<Long, SeekableUnderFileInputStream> removal) {
+          SeekableUnderFileInputStream inputStream = removal.getValue();
+          synchronized (mFileToInputStreamIds) {
+            if (mFileToInputStreamIds.containsKey(inputStream.getFilePath())) {
+              UfsInputStreamIds resources = mFileToInputStreamIds.get(inputStream.getFilePath());
+              synchronized (resources) {
+                // remove the key
+                resources.removeInUse(removal.getKey());
+                if (resources.removeAvailable(removal.getKey())) {
+                  // close the resource
+                  LOG.debug("Removed the under file input stream resource of {}", removal.getKey());
+                  try {
+                    inputStream.close();
+                  } catch (IOException e) {
+                    LOG.warn("Failed to close the input stream resource of file {} and resource id",
+                        inputStream.getFilePath(), removal.getKey());
+                  }
+                }
+                if (resources.isEmpty()) {
+                  // remove the resources entry
+                  mFileToInputStreamIds.remove(inputStream.getFilePath());
+                }
+              }
+            } else {
+              LOG.warn("Try to remove the resource entry of {} but not exists any more",
+                  removal.getKey());
+            }
+          }
+        }
+      };
+
+  /**
+   * Creates a new {@link UfsInputStreamManager}.
+   */
+  public UfsInputStreamManager() {
+    mFileToInputStreamIds = new HashMap<>();
+    mRemovalThreadPool = Executors.newFixedThreadPool(2);
+
+    mUnderFileInputStreamCache = CacheBuilder.newBuilder()
+        .maximumSize(Configuration.getInt(PropertyKey.WORKER_UFS_INSTREAM_CACHE_MAX_SIZE))
+        .expireAfterAccess(Configuration.getMs(PropertyKey.WORKER_UFS_INSTREAM_CACHE_EXPIRE_MS),
+            TimeUnit.MILLISECONDS)
+        .removalListener(RemovalListeners.asynchronous(mRemovalListener, mRemovalThreadPool))
+        .build();
+  }
+
+  /**
+   * Checks in an input stream. The input stream is closed if it's already expired.
+   *
+   * @param inputStream the input stream to check in
+   * @throws IOException when input stream fails to close
+   */
+  public void checkIn(InputStream inputStream) throws IOException {
```

Comment:
nit: can we rename this method to be `release` and the next to be `acquire`?
just to be more consistent with other resource management classes in this code base.

Reply:
sure